ParselyTracker/ParselyTracker.swift

```diff
@@ -282,27 +285,44 @@ public class Parsely {
         os_log("Stopping execution before background/inactive/terminate", log:OSLog.tracker, type:.info)
         hardShutdown()
 
-        self.backgroundFlushTask = UIApplication.shared.beginBackgroundTask(expirationHandler:{
-            self.endBackgroundFlushTask()
-        })
         os_log("Flushing queue in background", log:OSLog.tracker, type:.info)
         self.track.sendHeartbeats()
         self.flush()
-        self.endBackgroundFlushTask()
     }
     
     internal func hardShutdown() {
         pauseFlushTimer()
         track.pause()
     }
-    
-    private func endBackgroundFlushTask() {
-        UIApplication.shared.endBackgroundTask(backgroundFlushTask)
-        backgroundFlushTask = UIBackgroundTaskIdentifier.invalid
-    }
+
 }
 
 extension OSLog {
     private static var logger = "ParselyTracker"
     static let tracker = OSLog(subsystem: "ParselyTracker", category: "parsely_tracker")
 }
+
+private final class BackgroundTask {
+    static func begin() -> Self {
+        let task = Self()
+        task.taskIdentifier = UIApplication.shared.beginBackgroundTask {
+            task.end()
+        }
+        return task
+    }
+
+    private var taskIdentifier: UIBackgroundTaskIdentifier!
```

Comment:
Have you considered making this `Optional`? 

I guess given the small scope of this object, there's really no chance that it will be accessed before being set. However, the only overhead would be an assignment in the `guard` below

```swift
guard let taskIdentifier, taskIdentifier != .invalid else {
// ...
self.taskIdentifier = .invalid
```

Reply:
The previous version used `.invalid` as the default value. I prefer the approach you chose here because it is less ambiguous. Now the identifier will be `.invalid` only after the task finishes, not before is starts, too.